### PR TITLE
[minor]: Update median implementation

### PR DESCRIPTION
**Pull request by akurmustafa**
## Which issue does this PR close?

Improves the situation in [13550](https://github.com/apache/datafusion/issues/13550)

## Rationale for this change
While examining the median code to see where can be changed to improve performance. I recognized that while calculating 
 the median for slices with even number of entries, we call `select_nth_unstable_by` twice. First to bipartition slice according to index `n/2`. Then to bipartition first part by index `n/2-1`.  The second bipartitioning is indeed maximum calculation. 

This PR changes the second bipartitioning call with maximum. I am not sure, whether this will improve the performance but I think at the worst case performance will be same, and this version is slightly more readable.

## Are these changes tested?
Existing tests

### CODE REVIEW

File: datafusion/functions-aggregate/src/median.rs [hunk 1]
```diff
@@ -310,6 +310,18 @@ impl<T: ArrowNumericType> Accumulator for DistinctMedianAccumulator<T> {
     }
 }
 
+/// Get maximum entry in the slice,
+fn slice_max<T>(array: &[T::Native]) -> Option<T::Native>
+where
+    T: ArrowPrimitiveType,
+    T::Native: PartialOrd, // Ensure the type supports PartialOrd for comparison
+{
```
Comment (Dandandan) [inline datafusion/functions-aggregate/src/median.rs:318]: As `unwrap` is used, the code could be updated / simplified to work on arrays of size >= 1

Comment (akurmustafa) [reply datafusion/functions-aggregate/src/median.rs:318]: That makes sense to me, updated the implementation such that function doesn't return option in [0303c9](https://github.com/apache/datafusion/pull/13554/commits/0303c965466c3b443f85e799fc3d8392877dbd20)

File: datafusion/functions-aggregate/src/median.rs [hunk 2]
```diff
@@ -310,6 +310,19 @@ impl<T: ArrowNumericType> Accumulator for DistinctMedianAccumulator<T> {
...
+    // Make sure that, array is not empty.
+    debug_assert!(!array.is_empty());
+    array
+        .iter()
+        .fold(array[0], |max, &val| if max > val { max } else { val })
+}
```
[the remark this answers is not in the document]
Comment (akurmustafa) [reply datafusion/functions-aggregate/src/median.rs:324]: When I did this I got following error: 
```
^^^ the trait `std::cmp::Ord` is not implemented for `<T as arrow::array::ArrowPrimitiveType>::Native`, which is required by `&<T as arrow::array::ArrowPrimitiveType>::Native: std::cmp::Ord`
```
I guess, max expects `Ord` trait. However, `T::Native` implements `PartialOrd` trait.

[the remark this answers is not in the document]
Comment (akurmustafa) [reply datafusion/functions-aggregate/src/median.rs]: By the way, I could do this using `max_by` API.

[the remark this answers is not in the document]
Comment (Dandandan) [reply datafusion/functions-aggregate/src/median.rs]: Nice

[the remark this answers is not in the document]
Comment (Dandandan) [reply datafusion/functions-aggregate/src/median.rs]: I think we can use `Ord` though?

[the remark this answers is not in the document]
Comment (akurmustafa) [reply datafusion/functions-aggregate/src/median.rs]: I am not sure we can, at least I couldn't make it work. However, maybe I am missing something. By the way, following is the trait description for `ArrowPrimitiveType::Native`
```rust
pub trait ArrowNativeType:
    std::fmt::Debug + Send + Sync + Copy + PartialOrd + Default + private::Sealed + 'static
{
```
and signature of `max` is as follows:
```rust
    fn max(self) -> Option<Self::Item>
    where
        Self: Sized,
        Self::Item: Ord,
    {
        self.max_by(Ord::cmp)
    }
```

**Review by comphead — APPROVED** [review]
lgtm thanks @akurmustafa and @Dandandan for the review

**Pull request merged**